File: scripts/solar/solar-common/w3ts/handles/timer.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 3,["9"] = 3,["12"] = 8,["13"] = 8,["14"] = 8,["15"] = 8,["17"] = 11,["18"] = 8,["20"] = 8,["21"] = 14,["22"] = 14,["23"] = 14,["25"] = 10,["26"] = 42,["27"] = 43,["28"] = 44,["29"] = 42,["30"] = 50,["31"] = 51,["32"] = 52,["33"] = 50,["34"] = 58,["35"] = 59,["36"] = 60,["37"] = 58,["38"] = 68,["39"] = 69,["40"] = 70,["41"] = 68,["42"] = 76,["43"] = 77,["44"] = 76,["45"] = 80,["46"] = 81,["47"] = 80,["52"] = 22,["60"] = 29,["68"] = 36});
local ____exports = {}
local ____handle = require("solar.solar-common.w3ts.handles.handle")
local Handle = ____handle.Handle
---
-- @deprecated 请尽量使用BaseUtil.runLater 中心计时器替代 以减少内存泄漏
____exports.Timer = __TS__Class()
local Timer = ____exports.Timer
Timer.name = "Timer"
__TS__ClassExtends(Timer, Handle)
function Timer.prototype.____constructor(self)
    if Handle:initFromHandle() then
        Handle.prototype.____constructor(self)
    else
        Handle.prototype.____constructor(
            self,
            CreateTimer()
        )
    end
end
function Timer.prototype.destroy(self)
    DestroyTimer(self.handle)
    return self
end
function Timer.prototype.pause(self)
    PauseTimer(self.handle)
    return self
end
function Timer.prototype.resume(self)
    ResumeTimer(self.handle)
    return self
end
function Timer.prototype.start(self, timeout, periodic, handlerFunc)
    TimerStart(self.handle, timeout, periodic, handlerFunc)
    return self
end
function Timer.fromExpired(self)
    return self:fromHandle(GetExpiredTimer())
end
function Timer.fromHandle(self, handle)
    return self:getObject(handle)
end
__TS__SetDescriptor(
    Timer.prototype,
    "elapsed",
    {get = function(self)
        return TimerGetElapsed(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    Timer.prototype,
    "remaining",
    {get = function(self)
        return TimerGetRemaining(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    Timer.prototype,
    "timeout",
    {get = function(self)
        return TimerGetTimeout(self.handle)
    end},
    true
)
return ____exports

File: src/solar/solar-common/w3ts/handles/timer.ts
/** @noSelfInFile **/

import {Handle} from "./handle";

/**
 * @deprecated 请尽量使用BaseUtil.runLater 中心计时器替代 以减少内存泄漏
 */
export class Timer extends Handle<timer> {

    constructor() {
        if (Handle.initFromHandle()) {
            super();
        } else {
            super(CreateTimer());
        }
    }

    /**
     * 计时器经过的时间
     */
    public get elapsed(): number {
        return TimerGetElapsed(this.handle);
    }

    /**
     * 计时器剩余时间
     */
    public get remaining(): number {
        return TimerGetRemaining(this.handle);
    }

    /**
     * 计时器初始的时间
     */
    public get timeout(): number {
        return TimerGetTimeout(this.handle);
    }

    /**
     * 删除计时器 [R]
     */
    public destroy() {
        DestroyTimer(this.handle);
        return this;
    }

    /**
     * 暂停计时器 [R]
     */
    public pause() {
        PauseTimer(this.handle);
        return this;
    }

    /**
     * 恢复计时器 [R]
     */
    public resume() {
        ResumeTimer(this.handle);
        return this;
    }

    /**
     * 运行计时器 [C]
     * @deprecated 请尽量使用BaseUtil.runLater 中心计时器替代 以减少内存泄漏
     * (或者handlerFunc 使用命名函数 以减少内存泄漏)
     */
    public start(timeout: number, periodic: boolean, handlerFunc: () => void) {
        TimerStart(this.handle, timeout, periodic, handlerFunc);
        return this;
    }

    /**
     * 事件响应 - 计时器期满
     */
    public static fromExpired(): Timer {
        return this.fromHandle(GetExpiredTimer());
    }

    public static fromHandle(handle: timer): Timer {
        return this.getObject(handle);
    }
}
